### PR TITLE
tl: add build options

### PR DESCRIPTION
**Pull request by euclidianAce**
Adds luafilesystem dependency and the following config options
- build_dir
- source_dir
- files
- include
- exclude
- Original include -> include_dir

Hopefully these changes should remove the need for a Makefile for pure Teal projects (though probably not `tl` itself)

This contains some pretty big refactors to `tl` that I will try to outline here.
- Setting up the environment and preloaded modules has been abstracted into some methods for reusability
- `tl gen` and `tl check` now build up a source map for file names (`src_map: {string:string}`)
	- `tl run` will not build up a source map because it only relies on the args passed to it.
	- In the future there could probably be a `source_map` config option as well to just directly map inputs and outputs.
- There is some custom logic to recursively traverse sub-directories and build up the source map, This could probably be optimized in the future.

Some notes:

- Technically the lfs dependency can be made optional and only required to use things like `source_dir`. Technically `build_dir` can get by with some `os.execute("mkdir ...")` stuff, but that always feels hacky.
- This currently passes the test suite but I'm am unsure of how to write tests for these features, so help on that would be appreciated.
	- Especially with the `**/` pattern matching functions, there doesn't seem to be a way to extract local functions with the debug library, so putting them in a module seems like the only way to test them, but having a test suite for these would be helpful.

Constructive criticism/reviews are definitely appreciated, especially for testing this. :tada:

### CODE REVIEW

**Comment by euclidianAce** [comment]
(Should close #47, #31, and maybe #107?)

File: tl-dev-1.rockspec [hunk 1]
```diff
@@ -16,6 +16,9 @@ dependencies = {
 
    -- needed for the cli tool
    "argparse"
+
+   -- needed for --source-dir
+   "lfs"
```
Comment (hishamhm) [inline tl-dev-1.rockspec:21]: the dependency name for LuaRocks is called `luafilesystem` (I know...)

**Comment by hishamhm** [comment]
@pdesaulniers Could you take a look at this PR? :)

**Comment by pdesaulniers** [comment]
OK, I'm looking at this.

File: tl-dev-1.rockspec [hunk 2]
```diff
@@ -16,6 +16,9 @@ dependencies = {
 
    -- needed for the cli tool
    "argparse"
```
Comment (pdesaulniers) [inline tl-dev-1.rockspec:18]: There's a missing comma here (@hishamhm, I think CI should've caught this?)

**Comment by pdesaulniers** [comment]
> Especially with the **/ pattern matching functions, there doesn't seem to be a way to extract local functions with the debug library, so putting them in a module seems like the only way to test them, but having a test suite for these would be helpful.

So far, the CLI flags are tested by calling `tl` from the shell and looking at its output, like this: https://github.com/teal-language/tl/blob/5732126a259a67908d7017b3a150c00d5a97bfd5/spec/cli/include_spec.lua#L4-L16

Right now, it would be a bit difficult to write this kind of tests for `include` and `exclude`, because  `include` cannot be set with a command line option, and `tl` assumes that `tlconfig.lua` is contained inside the working directory.

So, for now, I would suggest to add some command line flags for `include` and `exclude`, and use these in the tests.

Otherwise, we could also add a `-p --project` flag for setting the path of `tlconfig.lua`, but this sounds a bit harder to manage (we might need one `tlconfig.lua` per test)

**Comment by pdesaulniers** [comment]
> (Should close #47, #31, and maybe #107?)

Yes for the first two, but I'm not sure about https://github.com/teal-language/tl/issues/107. I think that issue is more about merging all of the command line and `tlconfig.lua` stuff inside a single table. That way, I suppose we could get rid of the [valid_keys array](https://github.com/teal-language/tl/blob/5732126a259a67908d7017b3a150c00d5a97bfd5/tl#L39-L43), generate the [list of compiler options](https://github.com/teal-language/tl/blob/master/docs/compiler_options.md#list-of-compiler-options) programmatically, etc.

**Comment by euclidianAce** [comment]
> tl assumes that tlconfig.lua is contained inside the current working directory.

Since we're bringing in lfs for this feature, we could probably use `lfs.chdir` and build up some directories in /tmp (maybe copying the tl executable too?) and just throw some `helloworld.tl`s and the `tlconfig.lua`s that we want to test in there. When I get the chance I'll try to write something up.

File: tl-dev-1.rockspec (hunk 2 above)
Comment (hishamhm) [reply tl-dev-1.rockspec:18]: oh good catch, that's right

Pushing #178 to catch this

**Comment by hishamhm** [comment]
@euclidianAce sounds like a plan!

File: docs/compiler_options.md [hunk 3]
```diff
@@ -24,5 +24,81 @@ return {
 | Command line option | Config key | Type | Description |
 | --- | --- | --- | --- |
 | `-l --preload` | `preload_modules` | `{string}` | Execute the equivalent of `require('modulename')` before executing the tl script(s). |
-| `-I --include` | `include` | `{string}` | Prepend this directory to the module search path.
+| `-I --include-dir` | `include_dir` | `{string}` | Prepend this directory to the module search path.
 | `--skip-compat53` | | | Skip compat53 insertions.
+|| `include` | `{string}` | The set of files to compile/check. See below for details on patterns.
+|| `exclude` | `{string}` | The set of files to exclude. See below for details on patterns.
+| `-s --source-dir` | `source_dir` | `string` | Set the directory to be searched for files. `gen` will compile every .tl file in every subdirectory by default.
+| `-b --build-dir` | `build_dir` | `string` | Set the directory for generated files, mimicing the file structure of the source files.
```
Comment (pdesaulniers) [inline docs/compiler_options.md:32]: Minor nitpick: should be 'mimicking' instead of 'mimicing'

File: docs/compiler_options.md [hunk 4]
```diff
@@ -24,5 +24,81 @@ return {
...
+|| `files` | `{string}` | The names of files to be compiled. Does not accept patterns like `include`.
+
+### Include/Exclude patterns
+
+The `include` and `exclude` fields can have glob-like patterns in them:
+- `*`: Matches any number of characters (excluding directory separators)
+- `**/`: Matches any number subdirectories
+
+In addition, setting the `source_dir` has the effect of prepending `source_dir` to all patterns
```
Comment (pdesaulniers) [inline docs/compiler_options.md:41]: Minor nitpick: I would add a period at the end of this sentence.

**Comment by euclidianAce** [comment]
I've made some pretty big refactors to this now that I've been able to test it and I want to just outline them here:

- Previously `tl gen` was modified to build a project when run with no arguments. This has been reverted back and I've added `tl build` which has that behavior
- `tl build` has a new flag associated with it (inspired from Gentoo)
    - `-p --pretend`: don't compile, but type check and print what files would be written to
        - Similarly from Gentoo, an `--ask` flag could easily be implemented as well
- The current `tl gen` now works more like gcc:
    - files are generated in the current directory unless specified with `-o --output`
    - multiple files are concatenated into one when using -o
        - This feature practically comes for free, by just appending instead of writing. With how the current compiler works, source maps may not be all that hard to implement since there is no minifying.
        - Idk it that's enough to close #67 but its a start.
- `--skip-compat53` can be put in `tlconfig.lua` as `skip_compat53 = true`, unrelated, but I'm using this pr with some of my own projects and it was annoying to type each time.

The test coverage is pretty alright, but I still need to write more :P

**Comment by hishamhm** [comment]
@euclidianAce wow, impressive!!

Some commentary based on the above comment (haven't looked deeply into details of the code yet)

> Previously tl gen was modified to build a project when run with no arguments. This has been reverted back and I've added tl build which has that behavior

Nice!

> -p --pretend: don't compile, but type check and print what files would be written to 

I've seen this being called `--dry-run` in other programs, but I guess `--pretend` works just as well!

> multiple files are concatenated into one when using -o 

This bit I am a little concerned with. If multiple files just declare global functions, then this will work fine, but if each file declares a module, then mere concatenation of the outputs will not produce what I would consider the expected behavior: amalgamation scripts generally end up declaring multiple independent modules (like one would see on a big `.so` C module that declares multiple Lua modules). I would rather drop this feature for `tl gen` for now (producing an amalgamation script could be done as a separate feature in a separate PR — let's restrict PR scope to get this one merged sooner :) )

> --skip-compat53 can be put in tlconfig.lua as skip_compat53 = true, unrelated, but I'm using this pr with some of my own projects and it was annoying to type each time.

:+1: 

> The test coverage is pretty alright, but I still need to write more :P

If the coverage is alright I guess we can look into reviewing and merging this version. Let's just make sure the new features are covered in the docs.

Very cool work in this PR, improving the tooling is super important!

**Comment by euclidianAce** [comment]
For the time being I'll limit `--output` to asserting only one script argument. Didn't really think of the consequences of that one.

But this is pretty feature-complete for what I imagined the initial write up of this would be and while I still have some ideas, I think it would be feature creep. So over the next few days I'll try to improve test coverage and documentation to get this ready to merge. :tada:

**Comment by euclidianAce** [comment]
I'll leave this here as a reference for where I think I'm at and suggestions for where to go.
## TODO

- Test:
	- [x] `source_dir`
	- [x] `build_dir`
	- [ ] any file including option with relative vs non-relative paths
	- [x] `include+exclude` globbing
	- [x] basic interactions between `include+exclude`
	- [ ] `include+exclude` with multiple patterns
	- [x] basic interactions between `source_dir+build_dir`
	- [x] basic interactions between `source_dir+build_dir+include+exclude`
	- [ ] error handling for `source_dir`
		- i.e. `source_dir` not found: error out gracefully, probably mentioning source dir in some way
	- [ ] error handling for `build_dir`
		- there is some error handling, but not extensively tested yet
	- [x] `-o --output`
	- [ ] `-p --pretend --dry-run`
		- honestly this could just be a copy-paste of any instance where `util.run_mock_project` was used and just add `-p` and verify the output
	- [ ] `tl build`
		- any usage of `util.run_mock_project` is basically the test suite for `build`, but there could be more explicit tests
- Document:
	- [x] `-s --source_dir`
	- [x] `-b --build_dir`
	- [x] `files` tlconfig.lua entry
	- [x] `include+exclude`
		- [x] globbing patterns
	- [x] `-o --output`
	- [x] `-p --pretend --dry-run`
	- [x] `tl build`
	- [x] `skip_compat53` tlconfig.lua entry

In general, I think I need to write more tests where the expected output is an error.

File: tl [hunk 5]
```diff
@@ -245,92 +300,456 @@ for i, filename in ipairs(args["script"]) do
...
+      if not s_idx then
+         matches = false
+         break
+      end
+   end
+   return matches
+end
+local function matcher(str)
+   local chunks = {}
+   for piece in str_split(str, "**/") do
```
Comment (hishamhm) [inline tl:511]: since the rest of the code is using `path_separator`, should this use as well?

File: tl [hunk 6]
```diff
@@ -245,92 +300,456 @@ for i, filename in ipairs(args["script"]) do
...
+   local current_dir = self.dir
+   for dirname in str_split(path, path_separator) do
+      current_dir = current_dir[dirname]
+      if not current_dir then
+         return nil
       end
+   end
+   return current_dir
+end
+local function is_in(a, b) -- a is in b
```
Comment (hishamhm) [inline tl:636]: looks like this function is unused

**Comment by hishamhm** [comment]
I'd like to get this merged as soon as possible since it's getting too big to do a review in one go. :)

I took another look at the PR (not a super detailed line-by-line look, I admit! it's a lot of stuff!) and left a couple of comments, apart from those I think we can get it merged now

File: tl (hunk 6 above)
Comment (euclidianAce) [reply tl:636]: I think I had a plan at one point to print warnings when generating files in the `source_dir` when `build_dir` was present. :shrug:

**Comment by euclidianAce** [comment]
I'd say go ahead and merge this if you'd like. The most basic error cases have tests now and I've been using this since I made the pr on a bunch of tiny things and it hasn't had any trouble.

Here's a rundown for the changelog:
- `tl gen` now generates the file in the current directory, much like gcc and other tools
	- `-o --output` now lets you specify the output for a _single_ script
- `tl build` is a new command that will type check and compile a given set of files specified in `tlconfig.lua` of the current working directory or the command line. These new entries/flags are:
	- `-b --build-dir` `build_dir: string`: where generated files will end up
	- `-s --source-dir` `source_dir: string`: when provided, `build` will only include files from inside the given directory
	- `files: {string}`: a list of filenames to compile
	- `include` has been renamed to `--include-dir include_dir: {string}` to account for:
	- `include, exclude: {string}`: a list of patterns to find files to compile, what patterns are available are outlined in `docs/compiler_options.md`
- New flag `-p --pretend --dry-run`: don't write to any files, but type check and log what files would be written to.
- New config entry `skip_compat53: boolean` same as `--skip-compat53`

**Comment by euclidianAce** [comment]
Hey @hishamhm, any updates on getting this merged?

**Comment by hishamhm** [comment]
@euclidianAce hi, sorry for dropping the ball on this! I've been unable to put too much energy on Teal the last couple of weeks. I just merged it manually, squashing some of the commits together. Thank you for the nudge, this PR is a huge win for the project!

**Comment by euclidianAce** [comment]
Thanks, no worries!

**Pull request closed without merging**